graylog2-server/src/main/java/org/graylog2/rest/resources/system/field_types/FieldTypeChangeResource.java

```diff
@@ -0,0 +1,147 @@
+import org.graylog2.indexer.MongoIndexSet;
+import org.graylog2.indexer.indexset.CustomFieldMappings;
+import org.graylog2.indexer.indexset.IndexSetConfig;
+import org.graylog2.indexer.indexset.IndexSetService;
+import org.graylog2.indexer.indexset.MongoIndexSetService;
+import org.graylog2.shared.rest.resources.RestResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.util.Optional;
+import java.util.Set;
+
+import static org.graylog2.shared.rest.documentation.generator.Generator.CLOUD_VISIBLE;
+
+@RequiresAuthentication
+@Api(value = "System/FieldTypes", tags = {CLOUD_VISIBLE})
+@Path("/system/indices/field_types_change")
+@Produces(MediaType.APPLICATION_JSON)
+public class FieldTypeChangeResource extends RestResource {
+    private static final Logger LOG = LoggerFactory.getLogger(FieldTypeChangeResource.class);
+
+    private final IndexSetService indexSetService;
+    private final MongoIndexSet.Factory mongoIndexSetFactory;
+    private final MongoIndexSetService mongoIndexSetService;
+
+    @Inject
+    public FieldTypeChangeResource(final IndexSetService indexSetService,
+                                   final MongoIndexSet.Factory mongoIndexSetFactory,
+                                   final MongoIndexSetService mongoIndexSetService) {
+        this.indexSetService = indexSetService;
+        this.mongoIndexSetFactory = mongoIndexSetFactory;
+        this.mongoIndexSetService = mongoIndexSetService;
+    }
+
+    @PUT
+    @Path("{field_name}")
+    @Timed
+    @ApiOperation(value = "Change field type for certain index sets")
+    @ApiResponses(value = {
+            @ApiResponse(code = 403, message = "Unauthorized")
+    })
+    @NoAuditEvent("No audit for field type changes")
+    public Response changeFieldType(@ApiParam(name = "index_sets_ids", required = true)
+                                    @QueryParam("index_sets_ids") Set<String> indexSetsIds,
+                                    @ApiParam(name = "field_name", required = true)
+                                    @PathParam("field_name") String fieldName,
+                                    @ApiParam(name = "new_type", required = true)
+                                    @QueryParam("new_type") String newType,
+                                    @ApiParam(name = "rotate_immediately", defaultValue = "false")
+                                    @QueryParam("rotate_immediately") boolean rotateImmediately
+    ) {
+        checkPermissions(indexSetsIds);
+
+        //TODO: validation of fields
+
+        CustomFieldMappings.CustomFieldMapping customMapping = new CustomFieldMappings.CustomFieldMapping(fieldName, newType);
+
+        for (String indexSetId : indexSetsIds) {
```

Comment:
We should extract the actual code into some kind of a service, so the REST resource is not doing too much. This way we can also support reuse as well as supporting potential refactorings (where the actual flow is triggered from somewhere else - e.g. through an approval mechanism).